packages/devtools_app/lib/src/screens/profiler/cpu_profile_model.dart

```diff
@@ -357,9 +360,53 @@ class CpuProfileData {
         },
       });
     }
+
+    await _addProcessedUrlsToTraceObject(isolateId, traceObject);
+
     return CpuProfileData.parse(traceObject);
   }
 
+  /// Helper function for determining and updating the
+  /// [CpuProfileData.processedUrlKey] entry for each stack frame in
+  /// [traceObject].
+  ///
+  /// [isolateId] The id which is passed to the getIsolate RPC to load this
+  /// isolate.
+  /// [traceObject] A map where the cpu profile data for each frame is stored.
+  static Future<void> _addProcessedUrlsToTraceObject(
+    String isolateId,
+    Map<String, dynamic> traceObject,
+  ) async {
+    final stackFrames = traceObject[CpuProfileData.stackFramesKey]
+        .values
+        .cast<Map<String, dynamic>>();
+    final stackFramesWaitingOnPackageUri = <Map<String, dynamic>>[];
+    for (final stackFrameJson in stackFrames) {
+      final resolvedUrl =
+          stackFrameJson[CpuProfileData.resolvedUrlKey] as String?;
+      if (resolvedUrl != null && resolvedUrl.isNotEmpty) {
+        final packageUri =
+            serviceManager.resolvedUriManager.getPackageUri(resolvedUrl);
+        if (packageUri != null) {
+          stackFrameJson[CpuProfileData.processedUrlKey] = packageUri;
+        } else {
+          stackFramesWaitingOnPackageUri.add(stackFrameJson);
```

Comment:
instead of storing the stack frame json objects, just store the key instead. And then iterate through the keys below instead of all the stackFrameJson objects. You can iterate through `traceObject[CpuProfileData.stackFramesKey].entries` above to have access to both the key and the value in this for loop

Reply:
Since dart passes primitives by value then I think storing the keys as Strings would use a bunch more memory. I believe that the current approach is just an array of Map References, which would each be the size of the pointer to that map.